File: 351003/Razhnou/task361/publisher/src/main/java/by/bsuir/task361/publisher/controller/TweetV2Controller.java
package by.bsuir.task361.publisher.controller;

import by.bsuir.task361.publisher.dto.request.TweetRequestTo;
import by.bsuir.task361.publisher.dto.response.TweetResponseTo;
import by.bsuir.task361.publisher.service.SecuredTweetService;
import jakarta.validation.Valid;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.DeleteMapping;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.PutMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

import java.util.List;

@RestController
@RequestMapping("/api/v2.0/tweets")
public class TweetV2Controller {
    private final SecuredTweetService securedTweetService;

    public TweetV2Controller(SecuredTweetService securedTweetService) {
        this.securedTweetService = securedTweetService;
    }

    @PostMapping
    public ResponseEntity<TweetResponseTo> create(@Valid @RequestBody TweetRequestTo request) {
        return ResponseEntity.status(HttpStatus.CREATED).body(securedTweetService.create(request));
    }

    @GetMapping
    public List<TweetResponseTo> findAll() {
        return securedTweetService.findAll();
    }

    @GetMapping("/{id}")
    public TweetResponseTo findById(@PathVariable Long id) {
        return securedTweetService.findById(id);
    }

    @PutMapping
    public TweetResponseTo update(@Valid @RequestBody TweetRequestTo request) {
        return securedTweetService.update(request);
    }

    @DeleteMapping("/{id}")
    public ResponseEntity<Void> delete(@PathVariable Long id) {
        securedTweetService.delete(id);
        return ResponseEntity.noContent().build();
    }
}

File: 351002/Abrashin/publisher/src/main/kotlin/com/example/distcomp/controller/TweetV2Controller.kt
package com.example.distcomp.controller

import com.example.distcomp.dto.request.TweetRequestTo
import com.example.distcomp.dto.response.CreatorResponseTo
import com.example.distcomp.dto.response.NoteResponseTo
import com.example.distcomp.dto.response.StickerResponseTo
import com.example.distcomp.dto.response.TweetResponseTo
import com.example.distcomp.security.AuthorizationService
import com.example.distcomp.service.TweetService
import jakarta.validation.Valid
import org.springframework.http.HttpStatus
import org.springframework.web.bind.annotation.DeleteMapping
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.PatchMapping
import org.springframework.web.bind.annotation.PathVariable
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.PutMapping
import org.springframework.web.bind.annotation.RequestBody
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RequestParam
import org.springframework.web.bind.annotation.ResponseStatus
import org.springframework.web.bind.annotation.RestController

@RestController
@RequestMapping("/api/v2.0/tweets")
class TweetV2Controller(
    private val service: TweetService,
    private val authorizationService: AuthorizationService
) {
    @PostMapping
    @ResponseStatus(HttpStatus.CREATED)
    fun create(@Valid @RequestBody request: TweetRequestTo): TweetResponseTo {
        authorizationService.ensureCanManageTweetRequest(request)
        return service.create(request)
    }

    @GetMapping("/{id}")
    fun getById(@PathVariable id: Long): TweetResponseTo = service.getById(id)

    @GetMapping
    fun getAll(
        @RequestParam(defaultValue = "0") page: Int,
        @RequestParam(defaultValue = "10") size: Int,
        @RequestParam(defaultValue = "id,desc") sort: Array<String>
    ): List<TweetResponseTo> = service.getAll(page, size, sort)

    @PutMapping("/{id}")
    fun update(@PathVariable id: Long, @Valid @RequestBody request: TweetRequestTo): TweetResponseTo {
        authorizationService.ensureCanManageTweet(id, request.creatorId)
        return service.patch(id, request)
    }

    @PatchMapping("/{id}")
    fun patch(@PathVariable id: Long, @Valid @RequestBody request: TweetRequestTo): TweetResponseTo {
        authorizationService.ensureCanManageTweet(id, request.creatorId)
        return service.patch(id, request)
    }

    @GetMapping("/{id}/creator")
    fun getCreatorByTweetId(@PathVariable id: Long): CreatorResponseTo = service.getCreatorByTweetId(id)

    @GetMapping("/{id}/stickers")
    fun getStickersByTweetId(@PathVariable id: Long): List<StickerResponseTo> = service.getStickersByTweetId(id)

    @GetMapping("/{id}/notes")
    fun getNotesByTweetId(@PathVariable id: Long): List<NoteResponseTo> = service.getNotesByTweetId(id)

    @DeleteMapping("/{id}")
    @ResponseStatus(HttpStatus.NO_CONTENT)
    fun delete(@PathVariable id: Long) {
        authorizationService.ensureCanManageTweet(id)
        service.delete(id)
    }
}
